doc/design/libsvm_format_column.md

```diff
@@ -0,0 +1,112 @@
+    We can do the same implicit feature derivation to LibSVM format as CSV format.
+    
+In conclusion, we would choose to extend `NUMERIC` column to support LibSVM format column in this design.
+
+## Only support LibSVM data without label value
+
+There are 2 kinds of LibSVM data format: with label value and without label value. In this design, we only support LibSVM format without label value, it is because:
+
+- The actual label value used for training would be confused if the LibSVM data format column has label value.
+    - Users may choose the label in LibSVM data as the training label, but we have not supported to choose partial data of a column as training label yet.
+    - Users may choose another column as the training label, which is different with the label value in LibSVM data format column.
+- If the LibSVM data contains label value, users can easily split the label into a separate column. For example, the SQL statements in MySQL is like:
+    ```sql
+    CREATE TABLE split_table AS 
+    SELECT 
+    SUBSTRING_INDEX(libsvm_column_with_label, ' ', 1) label, 
+    TRIM(SUBSTR(libsvm_column_with_label, LOCATE(' ', libsvm_column_with_label))) libsvm_column_without_label
+    FROM source_table;
+    ```
+    
+Therefore, we choose to support LibSVM format without label only in this design, and suggest users to separate the label value using the SQL statement above
+if the LibSVM data contains label value.
+
+## Design
+
+### Add format field in `FieldDesc`
+
+Add a field named `format` in `FieldDesc`. It may be `csv`, `libsvm` or other data format we would support in the future.
+
+```go
+type FieldDesc struct {
+	Name      string
+	DType     int
+	Delimiter string
+	Shape     []int
+	IsSparse  bool
+	Format    string `json:"format"`    // indicates the data format
+	...
+}
+```
+
+### Implicit data format derivation in feature derivation stage
+
+In feature derivation stage, we can use a regular expression to check whether the input data format is LibSVM. 
+To avoid too many regular expression matching, this checking would be done only once when inferring the first row 
+of the fetched samples in feature derivation stage. That is to say (in pseudo code):
+
+```go
+func InferFeatureColumns() {
+    rows := FetchSamples()
+    rowCount := 0
+    for rows.Next() {
+        if rowCount == 0 {
+            format := inferDataFormat(rows.Value()) // Use regular expression to infer the data format
+            if format == "libsvm" {
+                // Fill FieldDesc info when the data format is LibSVM
+                // For example: FieldDesc.Format = "libsvm", FieldDesc.IsSparse = true, etc.
+                ...
+            }
+        }
+        rowCount ++
+    }
+}
+```
+
+### Convert the LibSVM data to sparse Tensor in Python data generator
+
+Whether the column is in LibSVM format is inferred in feature derivation stage, and the data format information would be 
```

Comment:
No passive voice. Always use active voice in technical writings.  Grammarly.com would remind you about the use of passive voice, so definitely use Grammarly.com anytime before submitting English writings. I would suggest you install the Chrome extension of Grammarly.com.

Reply:
Done.